File: package/rust/src/core/transport.rs
use std::sync::Arc;
use std::collections::HashMap;

use reqwest::Client;
use serde::de::DeserializeOwned;

use crate::errors::IdentityError;

use crate::types::IdentityClientConfig;

#[derive(Clone)]
pub struct Transport {
    base_url: String,
    client: Client,
    client_id: String,
}

impl Transport {
    pub fn new(base_url: String, client_id: String) -> Self {
        Self {
            base_url,
            client: Client::new(),
            client_id,
        }
    }

    async fn request<T: DeserializeOwned>(
        &self,
        endpoint: &str,
        method: &str,
        body: Option<&serde_json::Value>,
        access_token: Option<&str>,
    ) -> Result<T, IdentityError> {
        let url = format!("{}{}", self.base_url, endpoint);
        let mut req = match method {
            "GET" => self.client.get(&url),
            "POST" => self.client.post(&url),
            "PUT" => self.client.put(&url),
            "DELETE" => self.client.delete(&url),
            _ => self.client.get(&url),
        };

        req = req.header("Content-Type", "application/json");
        req = req.header("X-Client-ID", &self.client_id);
        if let Some(token) = access_token {
            req = req.bearer_auth(token);
        }
        if let Some(b) = body {
            req = req.json(b);
        }

        let resp = req.send().await.map_err(|_| IdentityError::NetworkError)?;
        if !resp.status().is_success() {
            // Try to parse error body for more details (best-effort)
            let status = resp.status().as_u16();
            let _ = (|| async move {
                let _d = resp.text().await; // ignore content for now
            })();
            // Map common statuses to errors
            return Err(match status {
                401 | 403 => IdentityError::AuthorizationError,
                419 => IdentityError::SessionExpiredError,
                423 => IdentityError::TOTPRequiredError,
                _ => IdentityError::ServerError,
            });
        }

        let data = resp.json::<T>().await.map_err(|_| IdentityError::ServerError)?;
        Ok(data)
    }

    pub async fn get<T: DeserializeOwned>(&self, endpoint: &str, access_token: Option<&str>) -> Result<T, IdentityError> {
        self.request(endpoint, "GET", None, access_token).await
    }

    pub async fn post<T: DeserializeOwned>(&self, endpoint: &str, data: Option<&serde_json::Value>, access_token: Option<&str>) -> Result<T, IdentityError> {
        self.request(endpoint, "POST", data, access_token).await
    }

    pub async fn put<T: DeserializeOwned>(&self, endpoint: &str, data: Option<&serde_json::Value>, access_token: Option<&str>) -> Result<T, IdentityError> {
        self.request(endpoint, "PUT", data, access_token).await
    }

    pub async fn delete<T: DeserializeOwned>(&self, endpoint: &str, access_token: Option<&str>) -> Result<T, IdentityError> {
        self.request(endpoint, "DELETE", None, access_token).await
    }
}

File: package/node/src/core/transport.ts
import type { FetchLike } from "../types";
import { createErrorFromResponse, NetworkError, ServerError } from "../errors";

const DEFAULT_MAX_RETRIES = 3;
const DEFAULT_RETRY_DELAY = 1000;

interface RequestOptions {
  method: string;
  headers: Record<string, string>;
  body?: string;
}

interface TransportConfig {
  baseUrl: string;
  fetcher: FetchLike;
  clientId: string;
  systemKey?: string;
  maxRetries?: number;
  retryDelay?: number;
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function isRetryableError(error: unknown): boolean {
  if (error instanceof NetworkError) return true;
  if (error instanceof ServerError) return true;
  return false;
}

class Transport {
  private config: TransportConfig;

  constructor(config: TransportConfig) {
    this.config = {
      maxRetries: DEFAULT_MAX_RETRIES,
      retryDelay: DEFAULT_RETRY_DELAY,
      ...config,
    };
  }

  private async request<T>(
    endpoint: string,
    options: RequestOptions,
    retries = this.config.maxRetries,
  ): Promise<T> {
    const url = `${this.config.baseUrl}${endpoint}`;

    try {
      const response = await this.config.fetcher(url, options);

      if (!response.ok) {
        let data: unknown;
        try {
          data = await response.json();
        } catch {
          data = {};
        }

        const error = createErrorFromResponse(response.status, data);

        if (isRetryableError(error) && retries! > 0) {
          await sleep(
            this.config.retryDelay! * (this.config.maxRetries! - retries! + 1),
          );
          return this.request<T>(endpoint, options, retries! - 1);
        }

        throw error;
      }

      return (await response.json()) as T;
    } catch (error) {
      if (error && typeof error === "object" && "code" in error) {
        throw error;
      }

      if (retries! > 0) {
        await sleep(
          this.config.retryDelay! * (this.config.maxRetries! - retries! + 1),
        );
        return this.request<T>(endpoint, options, retries! - 1);
      }

      throw new NetworkError("Network error occurred");
    }
  }

  async get<T>(endpoint: string, accessToken?: string): Promise<T> {
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      "X-Client-ID": this.config.clientId,
    };

    if (accessToken) {
      headers["Authorization"] = `Bearer ${accessToken}`;
    }

    if (this.config.systemKey) {
      headers["X-System-Key"] = this.config.systemKey;
    }

    return this.request<T>(endpoint, {
      method: "GET",
      headers,
    });
  }

  async post<T>(
    endpoint: string,
    data?: unknown,
    accessToken?: string,
    useSystemKeyAsAuth?: boolean,
  ): Promise<T> {
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      "X-Client-ID": this.config.clientId,
    };

    if (accessToken) {
      headers["Authorization"] = `Bearer ${accessToken}`;
    } else if (useSystemKeyAsAuth && this.config.systemKey) {
      headers["Authorization"] = `Bearer ${this.config.systemKey}`;
    }

    return this.request<T>(endpoint, {
      method: "POST",
      headers,
      body: data ? JSON.stringify(data) : undefined,
    });
  }

  async put<T>(
    endpoint: string,
    data?: unknown,
    accessToken?: string,
  ): Promise<T> {
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      "X-Client-ID": this.config.clientId,
    };

    if (accessToken) {
      headers["Authorization"] = `Bearer ${accessToken}`;
    }

    return this.request<T>(endpoint, {
      method: "PUT",
      headers,
      body: data ? JSON.stringify(data) : undefined,
    });
  }

  async delete<T>(endpoint: string, accessToken?: string): Promise<T> {
    const headers: Record<string, string> = {
      "Content-Type": "application/json",
      "X-Client-ID": this.config.clientId,
    };

    if (accessToken) {
      headers["Authorization"] = `Bearer ${accessToken}`;
    }

    if (this.config.systemKey) {
      headers["X-System-Key"] = this.config.systemKey;
    }

    return this.request<T>(endpoint, {
      method: "DELETE",
      headers,
    });
  }
}

export { Transport };
